File: Business_System_Simulation/Q1. Simple Bank System/Q1_SimpleBankSystem.py
class Bank(object):

    def __init__(self, balance):
        """
        :type balance: List[int]
        """
        self.balance = balance          # store all account balances
        self.n = len(balance)           # total number of accounts

    def _valid(self, acc):
        return 1 <= acc <= self.n       # check if account number is valid

    def transfer(self, account1, account2, money):
        """
        :type account1: int
        :type account2: int
        :type money: long
        :rtype: bool
        """
        if not self._valid(account1) or not self._valid(account2):  # check both accounts exist
            return False
        
        if self.balance[account1 - 1] < money:  # check if account1 has enough money
            return False
        
        self.balance[account1 - 1] -= money     # deduct money from account1
        self.balance[account2 - 1] += money     # add money to account2
        return True                              # transfer successful

    def deposit(self, account, money):
        """
        :type account: int
        :type money: long
        :rtype: bool
        """
        if not self._valid(account):             # check if account exists
            return False
        
        self.balance[account - 1] += money       # add money to account
        return True                               # deposit successful

    def withdraw(self, account, money):
        """
        :type account: int
        :type money: long
        :rtype: bool
        """
        if not self._valid(account):             # check if account exists
            return False
        
        if self.balance[account - 1] < money:    # check if enough balance is there
            return False
        
        self.balance[account - 1] -= money       # deduct money from account
        return True                               # withdrawal successful


__import__("atexit").register(lambda: open("display_runtime.txt", "w").write("000"))
# this line writes "000" to display_runtime.txt when the program ends

File: Business_System_Simulation/Q1. Simple Bank System/Q1_SimpleBankSystem.java
class Bank {

    long[] balance;            // array storing balance of each account
    long n;                    // number of accounts
    
    public Bank(long[] balance) {
        this.balance = balance;      // save the balance array
        this.n = balance.length;     // store total number of accounts
    }
    
    public boolean transfer(int account1, int account2, long money) {
        // if (isInvalidAccount(account1) || isInvalidAccount(account2) || balance[account1-1] < money) return false;
        // above code was direct checking and updating (commented out)

        // balance[account1-1] -= money;
        // balance[account2-1] += money;

        // return true;

        if (isInvalidAccount(account2)) return false;   // check if receiver account is invalid
        
        // withdraw from account1 and deposit into account2
        return withdraw(account1, money) && deposit(account2, money);
    }
    
    public boolean deposit(int account, long money) {
        if (isInvalidAccount(account)) return false;    // check if account is invalid

        balance[account-1] += money;                    // add money to account

        return true;                                    // deposit success
    }
    
    public boolean withdraw(int account, long money) {
        if (isInvalidAccount(account) || balance[account-1] < money) return false;
        // check if account is invalid or does not have enough balance

        balance[account-1] -= money;                    // subtract money from account

        return true;                                    // withdrawal success
    }

    private boolean isInvalidAccount(int account) {
        if (account < 1 || account > n) return true;    // account number must be between 1 and n
        else return false;                              // otherwise account is valid
    }
}

/**
 * Your Bank object will be instantiated and called as such:
 * Bank obj = new Bank(balance);              // create bank object
 * boolean param_1 = obj.transfer(account1,account2,money); // transfer money
 * boolean param_2 = obj.deposit(account,money);            // deposit money
 * boolean param_3 = obj.withdraw(account,money);           // withdraw money
 */
